pilot/pkg/serviceregistry/kube/controller/ambientindex.go

```diff
@@ -0,0 +1,864 @@
+		if action == workloadapi.RBACPolicyAction_ALLOW && anyNonEmpty(op.Hosts, op.NotHosts, op.Methods, op.NotMethods, op.Paths, op.NotPaths) {
+			// L7 policies are a deny
+			return nil
+		}
+		match := &workloadapi.RBACPolicyRuleMatch{
+			DestinationPorts:    stringToPort(op.Ports),
+			NotDestinationPorts: stringToPort(op.NotPorts),
+		}
+		toMatches = append(toMatches, match)
+	}
+	fromMatches := []*workloadapi.RBACPolicyRuleMatch{}
+	for _, from := range rule.From {
+		op := from.Source
+		if action == workloadapi.RBACPolicyAction_ALLOW && anyNonEmpty(op.RemoteIpBlocks, op.NotRemoteIpBlocks, op.RequestPrincipals, op.NotRequestPrincipals) {
+			// L7 policies are a deny
+			return nil
+		}
+		match := &workloadapi.RBACPolicyRuleMatch{
+			SourceIps:     stringToIP(op.IpBlocks),
+			NotSourceIps:  stringToIP(op.NotIpBlocks),
+			Namespaces:    stringToMatch(op.Namespaces),
+			NotNamespaces: stringToMatch(op.NotNamespaces),
+			Principals:    stringToMatch(op.Principals),
+			NotPrincipals: stringToMatch(op.NotPrincipals),
+		}
+		fromMatches = append(fromMatches, match)
+	}
+
+	rules := []*workloadapi.RBACPolicyRules{}
+	if len(toMatches) > 0 {
+		rules = append(rules, &workloadapi.RBACPolicyRules{Matches: toMatches})
+	}
+	if len(fromMatches) > 0 {
+		rules = append(rules, &workloadapi.RBACPolicyRules{Matches: fromMatches})
+	}
+	for _, when := range rule.When {
+		if action == workloadapi.RBACPolicyAction_ALLOW && !l4WhenAttributes.Contains(when.Key) {
+			// L7 policies are a deny
+			return nil
+		}
+		match := &workloadapi.RBACPolicyRuleMatch{
+			Namespaces:       whenMatch("source.namespace", when, false, stringToMatch),
+			Principals:       whenMatch("source.principal", when, false, stringToMatch),
+			SourceIps:        whenMatch("source.ip", when, false, stringToIP),
+			DestinationPorts: whenMatch("destination.port", when, false, stringToPort),
+			DestinationIps:   whenMatch("destination.ip", when, false, stringToIP),
+		}
+		rules = append(rules, &workloadapi.RBACPolicyRules{Matches: []*workloadapi.RBACPolicyRuleMatch{match}})
+	}
+	for _, when := range rule.When {
+		match := &workloadapi.RBACPolicyRuleMatch{
+			NotNamespaces:       whenMatch("source.namespace", when, true, stringToMatch),
+			NotPrincipals:       whenMatch("source.principal", when, true, stringToMatch),
+			NotSourceIps:        whenMatch("source.ip", when, true, stringToIP),
+			NotDestinationPorts: whenMatch("destination.port", when, true, stringToPort),
+			NotDestinationIps:   whenMatch("destination.ip", when, true, stringToIP),
+		}
+		rules = append(rules, &workloadapi.RBACPolicyRules{Matches: []*workloadapi.RBACPolicyRuleMatch{match}})
+	}
+	return rules
+}
+
+var l4WhenAttributes = sets.New(
+	"source.ip",
+	"source.namespace",
+	"source.principal",
+	"destination.ip",
+	"destination.port",
+)
+
+func whenMatch[T any](s string, when *v1beta1.Condition, invert bool, f func(v []string) []T) []T {
+	if when.Key != s {
+		return nil
+	}
+	if invert {
```

Comment:
Is this backwards?

Reply:
Looks like it indeed should take the not values for the invert